### PR TITLE
Support .gitattributes export-ignore in distribution archives

### PR DESCRIPTION
**Pull request by jberdine**
Parse .gitattributes files and exclude paths marked with the export-ignore
attribute from distribution tarballs. This is the mechanism used by
git-archive to exclude files, allowing projects to exclude dev-only files
like dune-workspace from releases.

Supported patterns:
- Exact matches: dune-workspace
- Directory patterns: .github/**
- Glob patterns: *.log, test_*, file?.txt
- Double star: **/build, src/**/test.ml

Paths are normalized before matching (handles ./ and ../).

Not supported:
- Escaped patterns (\\! for literal \!)
- Quoted patterns ("a b" for patterns with spaces)
- Case insensitivity (core.ignorecase)
- Negation patterns (\!pattern)
- Subdirectory .gitattributes files

Testing:
- Unit tests comparing behavior against git check-attr
- Tests cover pattern matching and parsing edge cases
- .gitattributes content is generated from test cases to ensure sync
- Archive integration test verifies end-to-end exclusion without requiring git

Signed-off-by: Josh Berdine <josh@berdine.net>

### CODE REVIEW

File: lib/archive.ml [hunk 1]
```diff
@@ -117,33 +117,45 @@ module Tar = struct
     String.concat (List.rev (end_of_file :: t))
 end
 
-let path_set_of_dir dir ~exclude_paths =
-  let not_excluded p = Ok (not (Fpath.Set.mem (Fpath.base p) exclude_paths)) in
+let path_set_of_dir dir ~exclude_paths ~export_ignore =
+  let not_excluded p =
```
Comment (Leonidas-from-XIV) [inline lib/archive.ml:121]: To avoid the negation, maybe rename it to `included`. It's kind of hard to wrap one's head around the double negated name and then the negation in the code.

File: lib/archive.ml [hunk 2]
```diff
@@ -117,33 +117,45 @@ module Tar = struct
...
     in
-    Logs.info (fun m -> m "Archiving %a" Fpath.pp fname);
     tar >>= fun tar ->
     OS.Dir.exists file >>= function
-    | true -> Tar.add tar fname ~mode:0o775 ~mtime `Dir
+    | true ->
+        (* Skip directories - they will be created implicitly when their
+           contents are added. This ensures that directories whose contents
+           are excluded via export-ignore patterns don't appear as empty
+           directories in the archive. *)
```
Comment (Leonidas-from-XIV) [inline lib/archive.ml:149]: But isn't this what one would expect to do? If I exclude a directory, I would expect the directory to show up, but not its contents.

Git does this but its more to do with the fact that Git doesn't track directories at all, so an empty directory is unrepresentable. But `dune-release` doesn't have this issue.

File: lib/archive.mli [hunk 3]
```diff
@@ -13,16 +13,19 @@ open Bos_setup
...
-(** [tar dir ~exclude_paths ~root ~mtime] is a (us)tar archive that contains the
-    file hierarchy [dir] except the relative hierarchies present in
-    [exclude_paths]. In the archive, members of [dir] are rerooted at [root] and
-    sorted according to {!Fpath.compare}. They have their modification time set
-    to [mtime] and their file permissions are [0o775] for directories and files
-    executable by the user and [0o664] for other files. No other file metadata
-    is preserved.
+(** [tar dir ~exclude_paths ~export_ignore ~root ~mtime] is a (us)tar archive
+    that contains the file hierarchy [dir] except:
+    - relative hierarchies present in [exclude_paths] (basename matching)
```
Comment (Leonidas-from-XIV) [inline lib/archive.mli:22]: I wonder if it wouldn't make more sense to just translate `exclude_paths` to a `Gitattributes.pattern` and use the more generic mechanism. That way there's only one way to ignore paths.

File: lib/gitattributes.ml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  | Prefix of string
+      (** Pattern like [dir/**] that matches everything under a directory, but
+          not the directory itself. *)
+  | Glob of Re.re  (** Compiled glob pattern. *)
+
+(** [glob_to_re pattern] is a compiled regex for glob [pattern]. Supports [*]
+    (any chars except /), [?] (single char except /), and [**] (any path
+    segments, but only when adjacent to /). *)
+let glob_to_re pattern =
+  let buf = Buffer.create (String.length pattern * 2) in
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:14]: Why is the initial size `pattern * 2`?

File: lib/gitattributes.ml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  Buffer.add_char buf '^';
+  let len = String.length pattern in
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:16]: Nitpick but the length of the pattern is determined twice, so could be pulled up in the function an reused.

File: lib/gitattributes.ml [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  let rec loop i =
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:17]: I wonder if it wouldn't make sense to use a `string list` and then `List.rev |> String.concat` instead of a mutable buffer here.

File: lib/gitattributes.ml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+          (* Escape regex metacharacters *)
+          Buffer.add_char buf '\\';
+          Buffer.add_char buf c;
+          loop (i + 1)
+      | _ ->
+          (* Literal character *)
+          Buffer.add_char buf c;
+          loop (i + 1)
+  in
+  loop 0;
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:58]: Maybe it could be more maintainable to use `ocamllex` for this instead of hand-writing a tokenizer where the tokens are regex snippets.

File: lib/gitattributes.ml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      (* Match everything under the directory, but not the directory itself *)
+      String.is_prefix ~affix:(prefix ^ "/") path_str
+  | Glob re ->
+      (* Match against full path or basename for patterns like *.log *)
+      Re.execp re path_str || Re.execp re basename
+
+let parse_export_ignore content =
+  (* Strip UTF-8 BOM if present at start of file *)
+  let content =
+    if String.is_prefix ~affix:"\xef\xbb\xbf" content then
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:105]: I would suggest putting `"\xef\xbb\xbf"` into its own binding.

File: lib/gitattributes.ml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      String.Sub.to_string (String.sub ~start:3 content)
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.ml:106]: And use `String.length utf8_bom` here.

File: lib/gitattributes.mli [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    Parses [.gitattributes] files and extracts patterns marked with the
+    [export-ignore] attribute. These patterns can be used to exclude files from
+    distribution archives. *)
+
+open Bos_setup
+
+(** {1 Patterns} *)
+
+type pattern
```
Comment (Leonidas-from-XIV) [inline lib/gitattributes.mli:11]:
```suggestion
type t
```

File: tests/lib/test_gitattributes.ml [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    integration without requiring git. *)
+let test_archive_export_ignore () =
+  let ( >>= ) = Result.bind in
+  let result =
+    Bos.OS.Dir.tmp "archive-test-%s" >>= fun dir ->
+    (* Create test file structure *)
+    let files =
+      [
+        ("CHANGES.md", "changes");
+        ("foo.opam", "opam");
```
Comment (Leonidas-from-XIV) [inline tests/lib/test_gitattributes.ml:24]:
```suggestion
        ("foo.opam", {|opam-version: "2.0"|});
```

File: tests/lib/test_gitattributes.ml [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    in
+    let create_file (path, content) =
+      let fpath = Fpath.(dir // v path) in
+      let parent = Fpath.parent fpath in
+      Bos.OS.Dir.create ~path:true parent >>= fun _ ->
+      Bos.OS.File.write fpath content
+    in
+    List.fold_left
+      (fun acc file -> acc >>= fun () -> create_file file)
+      (Ok ()) files
```
Comment (Leonidas-from-XIV) [inline tests/lib/test_gitattributes.ml:44]: I think `List.iter` followed by some kind of `Result.List.all_unit` (or the like) would be a bit easier to understand but it's of minor importance.

File: README.md [hunk 13]
```diff
@@ -110,6 +110,31 @@ The full documentation of this command is available with
...
+```
+
+**Supported patterns:**
+- Exact filenames: `dune-workspace`
+- Directory patterns: `.github/**` (matches all files under `.github/`)
+- Glob patterns: `*.log`, `test_*`, `file?.txt`
+- Double star in path: `**/build`, `src/**/test.ml`
+- Path normalization: handles `./` and `../` in paths
+
+**Not supported:**
```
Comment (Leonidas-from-XIV) [inline README.md:130]: I think it would be good to state what "not supported" means in practice. Will it error? Will it ignore the pattern?

**Review by Leonidas-from-XIV — COMMENTED** [review]
Very nice PR, the extensive test coverage is definitely a highlight and good to see as the glob to regex code is somewhat hairy.

I have some suggestions for improvements here.

File: lib/archive.ml (hunk 1 above)
Comment (jberdine) [reply lib/archive.ml]: Renamed. My thinking was that conceptually what is being done is "excluding" and so kept that base term, but fair point about double-negation.

File: lib/archive.ml (hunk 2 above)
Comment (jberdine) [reply lib/archive.ml:149]: My thinking is that the context here is reading a `.gitattributes` file, and so the expectation is to interpret it the way that `git archive` does (as much as possible). As a user, if some workflows used `git archive` and some `dune release`, and I had to adjust the `.gitattributes` file because they interpreted it differently, I would assume that `dune` was wrong since the name `.gitattributes` indicates that `git` "owns" the format.

File: lib/archive.mli (hunk 3 above)
Comment (jberdine) [reply lib/archive.mli:22]: That could make sense. My worry is that `exclude_paths` excludes based on basename, so at any depth, while `Gitattributes.pattern`s are path-anchored. Translating `exclude_paths` could have a semantic change if not done exactly right. So I think such a change ought to be in a separate PR, and I would want to have a much better idea for a good way to test it than I have at the moment.

File: lib/gitattributes.ml (hunk 4 above)
Comment (jberdine) [reply lib/gitattributes.ml]: Dropped the `* 2`. It is a guess, based on the observation that each char of the pattern might expand into 2 regex chars, but the buffer grows fine so it is simpler to just skip the `* 2`.

File: lib/gitattributes.ml (hunk 5 above)
Comment (jberdine) [reply lib/gitattributes.ml:14]: Fixed.

File: lib/gitattributes.ml (hunk 6 above)
Comment (jberdine) [reply lib/gitattributes.ml:17]: That could be done. But I think that using a Buffer for building a string char-by-char is idiomatic, and the mutable state is encapsulated within a single function.

File: lib/gitattributes.ml (hunk 7 above)
Comment (jberdine) [reply lib/gitattributes.ml:58]: Maybe, but I don't know. It would add a build dependency and generated file, and I don't expect that the syntax of globs is likely to change often. It could just as easily be done later, but I don't know if I will have time to work on that soon.

File: lib/gitattributes.ml (hunk 8 above)
Comment (jberdine) [reply lib/gitattributes.ml]: Done.

File: lib/gitattributes.ml (hunk 9 above)
Comment (jberdine) [reply lib/gitattributes.ml]: Done.

File: lib/gitattributes.mli (hunk 10 above)
Comment (jberdine) [reply lib/gitattributes.mli]: Done.

File: tests/lib/test_gitattributes.ml (hunk 11 above)
Comment (jberdine) [reply tests/lib/test_gitattributes.ml]: Done.

File: tests/lib/test_gitattributes.ml (hunk 12 above)
Comment (jberdine) [reply tests/lib/test_gitattributes.ml:44]: Perhaps. I like the way the current formulation uses `>>=`/`Result.bind` to chain the results just like the rest of the code.

File: README.md (hunk 13 above)
Comment (jberdine) [reply README.md]: Good point. I have pushed a commit that detects the unsupported pattern syntax, and updated the README to distinguish such cases (that are skipped with a warning) and other limitations (that are silent).
